File: app/src/main/java/com/android/nikhil/analyse/database/TextsContentProvider.java
package com.android.nikhil.analyse.database;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

/**
 * Created by NIKHIL on 02-01-2018.
 */

public class TextsContentProvider extends ContentProvider {

    private static final String TAG = "TextsContentProvider";
    private static UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
    private TextHelper mHelper;
    private Context context;

    private static final int CODE_ITEM = 1000;
    private static final int CODE_DIR = 1001;

    static {
        matcher.addURI(ContractClass.AUTHORITY, ContractClass.TextProvider.PATH_ITEM, CODE_ITEM);
        matcher.addURI(ContractClass.AUTHORITY, ContractClass.TextProvider.PATH_DIR, CODE_DIR);
    }

    @Override
    public boolean onCreate() {
        context = getContext();
        mHelper = new TextHelper(context);
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {

        Cursor cursor;
        SQLiteDatabase database = mHelper.getReadableDatabase();

        int code = matcher.match(uri);

        switch (code) {
            case CODE_DIR:
                cursor = database.query(ContractClass.TextProvider.TABLE_NAME,
                        projection, null, null, null, null, null);
                break;
            case CODE_ITEM:
                selection = ContractClass.TextProvider.COLUMN_ID + "=?";
                selectionArgs = new String[]{String.valueOf(ContentUris.parseId(uri))};
                cursor = database.query(ContractClass.TextProvider.TABLE_NAME,
                        projection, selection, selectionArgs, null, null, null);
                break;
            default:
                throw new IllegalArgumentException("Invalid uri : " + uri.toString());
        }
        cursor.setNotificationUri(context.getContentResolver(), uri);
        Log.d(TAG, "query: data queried.");
        return cursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        int code = matcher.match(uri);
        switch (code) {
            case CODE_DIR:
                return ContractClass.CONTENT_DIR_TYPE;
            case CODE_ITEM:
                return ContractClass.CONTENT_ITEM_TYPE;
            default:
                throw new IllegalArgumentException("Invalid Uri : " + uri.toString());
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        SQLiteDatabase database = mHelper.getWritableDatabase();
        Long id = database.insert(ContractClass.TextProvider.TABLE_NAME, null, values);
        if (id != -1)
            return ContentUris.withAppendedId(Uri.withAppendedPath(ContractClass.BASE_URI, ContractClass.TextProvider.PATH_DIR), id);
        else
            throw new IllegalArgumentException("Failed to insert for uri:" + uri.toString());
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        SQLiteDatabase database = mHelper.getWritableDatabase();
        int code = matcher.match(uri);
        int row;
        switch (code) {
            case CODE_DIR:
                row = database.delete(ContractClass.TextProvider.TABLE_NAME, selection, selectionArgs);
                break;
            case CODE_ITEM:
                selection = ContractClass.TextProvider.COLUMN_ID + "=?";
                selectionArgs = new String[]{String.valueOf(ContentUris.parseId(uri))};
                row = database.delete(ContractClass.TextProvider.TABLE_NAME, selection, selectionArgs);
                break;
            default:
                throw new IllegalArgumentException("Failed to delete data for uri:" + uri.toString());
        }
        return row;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        SQLiteDatabase database = mHelper.getWritableDatabase();
        int code = matcher.match(uri);
        int row;
        switch (code) {
            case CODE_DIR:
                row = database.update(ContractClass.TextProvider.TABLE_NAME, values, selection, selectionArgs);
                break;
            case CODE_ITEM:
                selection = ContractClass.TextProvider.COLUMN_ID + "=?";
                selectionArgs = new String[]{String.valueOf(ContentUris.parseId(uri))};
                row = database.update(ContractClass.TextProvider.TABLE_NAME, values, selection, selectionArgs);
                break;
            default:
                throw new IllegalArgumentException("Failed to update for uri:" + uri.toString());
        }
        return row;
    }
}

File: app/src/main/java/com/android/nikhil/analyse/database/TextsContentProvider.kt
package com.android.nikhil.analyse.database

import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.Context
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.net.Uri
import android.util.Log

/**
 * Created by NIKHIL on 02-01-2018.
 */

class TextsContentProvider : ContentProvider() {
  private var mHelper: TextHelper? = null

  override fun onCreate(): Boolean {
    mHelper = TextHelper(context)
    return true
  }

  override fun query(
    uri: Uri,
    projection: Array<String>?,
    selection: String?,
    selectionArgs: Array<String>?,
    sortOrder: String?
  ): Cursor? {
    var selection = selection
    var selectionArgs = selectionArgs

    val cursor: Cursor
    val database = mHelper!!.readableDatabase

    val code = matcher.match(uri)

    when (code) {
      CODE_DIR -> cursor = database.query(
          ContractClass.TextProvider.TABLE_NAME,
          projection, null, null, null, null, null
      )
      CODE_ITEM -> {
        selection = ContractClass.TextProvider.COLUMN_ID + "=?"
        selectionArgs = arrayOf(ContentUris.parseId(uri).toString())
        cursor = database.query(
            ContractClass.TextProvider.TABLE_NAME,
            projection, selection, selectionArgs, null, null, null
        )
      }
      else -> throw IllegalArgumentException("Invalid uri : " + uri.toString())
    }
    cursor.setNotificationUri(context!!.contentResolver, uri)
    Log.d(TAG, "query: data queried.")
    return cursor
  }

  override fun getType(uri: Uri): String? {
    val code = matcher.match(uri)
    return when (code) {
      CODE_DIR -> ContractClass.CONTENT_DIR_TYPE
      CODE_ITEM -> ContractClass.CONTENT_ITEM_TYPE
      else -> throw IllegalArgumentException("Invalid Uri : " + uri.toString())
    }
  }

  override fun insert(
    uri: Uri,
    values: ContentValues?
  ): Uri? {
    val database = mHelper!!.writableDatabase
    val id = database.insert(ContractClass.TextProvider.TABLE_NAME, null, values)
    return if (id.toInt() != -1)
      ContentUris.withAppendedId(Uri.withAppendedPath(ContractClass.BASE_URI, ContractClass.TextProvider.PATH_DIR), id)
    else
      throw IllegalArgumentException("Failed to insert for uri:" + uri.toString())
  }

  override fun delete(
    uri: Uri,
    selection: String?,
    selectionArgs: Array<String>?
  ): Int {
    var selection = selection
    var selectionArgs = selectionArgs
    val database = mHelper!!.writableDatabase
    val code = matcher.match(uri)
    val row: Int
    when (code) {
      CODE_DIR -> row = database.delete(ContractClass.TextProvider.TABLE_NAME, selection, selectionArgs)
      CODE_ITEM -> {
        selection = ContractClass.TextProvider.COLUMN_ID + "=?"
        selectionArgs = arrayOf(ContentUris.parseId(uri).toString())
        row = database.delete(ContractClass.TextProvider.TABLE_NAME, selection, selectionArgs)
      }
      else -> throw IllegalArgumentException("Failed to delete data for uri:" + uri.toString())
    }
    return row
  }

  override fun update(
    uri: Uri,
    values: ContentValues?,
    selection: String?,
    selectionArgs: Array<String>?
  ): Int {
    var selection = selection
    var selectionArgs = selectionArgs
    val database = mHelper!!.writableDatabase
    val code = matcher.match(uri)
    val row: Int
    when (code) {
      CODE_DIR -> row = database.update(ContractClass.TextProvider.TABLE_NAME, values, selection, selectionArgs)
      CODE_ITEM -> {
        selection = ContractClass.TextProvider.COLUMN_ID + "=?"
        selectionArgs = arrayOf(ContentUris.parseId(uri).toString())
        row = database.update(ContractClass.TextProvider.TABLE_NAME, values, selection, selectionArgs)
      }
      else -> throw IllegalArgumentException("Failed to update for uri:" + uri.toString())
    }
    return row
  }

  companion object {

    private const val TAG = "TextsContentProvider"
    private val matcher = UriMatcher(UriMatcher.NO_MATCH)

    private const val CODE_ITEM = 1000
    private const val CODE_DIR = 1001

    init {
      matcher.addURI(ContractClass.AUTHORITY, ContractClass.TextProvider.PATH_ITEM, CODE_ITEM)
      matcher.addURI(ContractClass.AUTHORITY, ContractClass.TextProvider.PATH_DIR, CODE_DIR)
    }
  }
}
